### PR TITLE
quickbrick tests and bug fixes

### LINKED ISSUE #96: Add travis tests of quickgen and quickbrick
These scripts are currently not tested by travis or included in coverage reports.

I believe the main obstacle is that any test would need to read the following files that currently live in the `desimodel` svn package:
```
Loaded 65000 rows from .../desimodel/data/spectra/spec-sky.dat with args {'format': 'ascii'}
Loaded 65000 rows from .../desimodel/data/spectra/spec-sky-bright.dat with args {'format': 'ascii'}
Loaded 65000 rows from .../desimodel/data/spectra/spec-sky-grey.dat with args {'format': 'ascii'}
Loaded 65000 rows from .../desimodel/data/spectra/ZenithExtinction-KPNO.dat with args {'format': 'ascii'}
Loaded 667 rows from .../desimodel/data/sky/solarspec.txt with args {'format': 'ascii.basic'}
Loaded 4233 rows from .../desimodel/data/specpsf/psf-quicksim.fits with args {'hdu': 'QUICKSIM-R'}
Loaded 22411 rows from .../desimodel/data/throughput/thru-r.fits with args {'hdu': 'THROUGHPUT'}
Loaded 4761 rows from .../desimodel/data/specpsf/psf-quicksim.fits with args {'hdu': 'QUICKSIM-B'}
Loaded 24651 rows from .../desimodel/data/throughput/thru-b.fits with args {'hdu': 'THROUGHPUT'}
Loaded 4799 rows from .../desimodel/data/specpsf/psf-quicksim.fits with args {'hdu': 'QUICKSIM-Z'}
Loaded 25531 rows from .../desimodel/data/throughput/thru-z.fits with args {'hdu': 'THROUGHPUT'}
Loaded 27 rows from .../desimodel/data/throughput/fiberloss-perfect.dat with args {'format': 'ascii'}
Loaded 27 rows from .../desimodel/data/throughput/fiberloss-star.dat with args {'format': 'ascii'}
Loaded 27 rows from .../desimodel/data/throughput/fiberloss-lrg.dat with args {'format': 'ascii'}
Loaded 27 rows from .../desimodel/data/throughput/fiberloss-qso.dat with args {'format': 'ascii'}
Loaded 27 rows from .../desimodel/data/throughput/fiberloss-sky.dat with args {'format': 'ascii'}
Loaded 27 rows from .../desimodel/data/throughput/fiberloss-elg.dat with args {'format': 'ascii'}
Loaded 65000 rows from .../desimodel/data/spectra/spec-ABmag22.0.dat with args {'format': 'ascii'}
```
(The list above was produced by running `quickspecsim -c desi --verbose`.)

Another problem is that `desisim` does not have a formal dependency on `specsim` that I can see.  My guess is that we need to updated `requirements.txt` and `etc/travis_env_common.sh` to fix this.

Finally, new unit tests need to be created that invoke quickgen and quickbrick.  I am not sure how to do this for standalone scripts under `bin/`, but it would be straightforward if we adopt the entrypoints mechanism recommended [here](http://astropy.readthedocs.org/en/latest/development/scripts.html).

### LINKED ISSUE #161: Update quickbrick and quickgen to specsim v0.5
The specsim exposure time parameter moved from `qsim.instrument` to the new`qsim.observation` in v0.5, but quickbrick and quickgen are still setting the value in `qsim.instrument` (which does not raise any exception, since python allows you to add attributes to an object on the fly).  The end result is that the exposure time is always fixed at 1000s.

This issue is to fix this problem, test the fix, and check for any other gotchas with the transition to specsim v0.5.

Thanks to @moustakas for flagging this!

### LINKED ISSUE #162: quickbrick does not run
The following command fails in the master branch:
```
quickbrick -b test1 --objtype ELG -n 1
```
The output (under python 2.7) is:
```
INFO:io.py:612:read_basis_templates: Reading /Data/DESI/basis_templates/v2.1/elg_templates_v2.0.fits
Traceback (most recent call last):
  File "/Users/david/anaconda/bin/quickbrick", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/Users/david/Cosmo/DESI/code/desisim/bin/quickbrick", line 16, in <module>
    quickbrick.main(args)
  File "/Users/david/Cosmo/DESI/code/desisim/py/desisim/scripts/quickbrick.py", line 232, in main
    meta.add_column(Column(true_objtype, dtype=(str, 10), name='OBJTYPE'))
  File "/Users/david/anaconda/lib/python2.7/site-packages/astropy/table/table.py", line 1387, in add_column
    self.add_columns([col], [index], rename_duplicate=rename_duplicate)
  File "/Users/david/anaconda/lib/python2.7/site-packages/astropy/table/table.py", line 1495, in add_columns
    self._init_from_cols(newcols)
  File "/Users/david/anaconda/lib/python2.7/site-packages/astropy/table/table.py", line 697, in _init_from_cols
    self._make_table_from_cols(self, newcols)
  File "/Users/david/anaconda/lib/python2.7/site-packages/astropy/table/table.py", line 729, in _make_table_from_cols
    raise ValueError('Duplicate column names')
ValueError: Duplicate column names
```
It looks like the code really is adding a duplicate OBJTYPE column. [According to blame](https://github.com/desihub/desisim/blame/master/py/desisim/scripts/quickbrick.py#L232) the offending line was last edited by @sbailey as part of his Py3 updates.

(This is a followup to #161 but is not related to specsim.)

### PR DESCRIPTION
**Pull request by sbailey**
This PR
 * adds tests for quickbrick
 * fixes #161 (updating how exposure time is set with specsim v0.5)
 * fixes #162 (quickbrick crash)
 * closes #96 (tests for quickgen and quickbrick)

It also fixes some bugs that weren't previously known and were caught by the new unit tests:
 * --seed option wasn't working
 * truth table output was corrupted using py3

However:
  * The tests won't work on Travis until desihub/specsim#43 is addressed to reduce the specsim memory footprint or another workaround is found.  I've flagged these with `@unittest.skipIf('TRAVIS_JOB_ID' in os.environ)` wrappers.  Note that this also impacts the quickgen tests -- you can run them on a laptop with sufficient memory, but not on Travis.  In the meantime, we also won't get accurate coverage reports on coveralls.io for these.
  * The tests are slow (2.5 minutes for just the quickbrick tests).  Although quicksim is fast for N>>1 spectra in a single shot, generating the templates on the fly is not, and quicksim has a fairly large overhead that we encounter multiple times when comparing things like "did making the moon brighter make the spectra noiser?"
  * I'm not sure what the original intent of OBJTYPE vs. TRUE_OBJTYPE was for the truth table.  What I have in there currently at least runs, but it may need further modification.

Related: apologies for the noise I generated on issue #161 where I kept typing "specsim" when I meant "desisim".  @dkirkby and I meant the same thing but I kept typing the wrong thing and misreading the right thing that he was typing...

### CODE REVIEW

**Comment by sbailey** [comment]
Any comments / complaints / suggestions?  Or "hold off I'll review this next week..."?

**Comment by moustakas** [comment]
I hope to take a look this weekend!

**Comment by sbailey** [comment]
Last call for comments.  This is becoming a blocking factor for other PR, e.g. #169 which is failing tests that are fixed in this PR.  I'll merge this afternoon with or without comments and we can retroactively fix stuff if that was a huge mistake.

File: py/desisim/test/test_quickbrick.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        t2 = fits.getdata('{}/brick-b-test2.fits'.format(self.testdir), '_TRUTH')
+        self.assertTrue(np.all(t1a['TRUE_OBJTYPE'] == t1b['TRUE_OBJTYPE']))
+        self.assertTrue(np.all(t1b['TRUEZ'] == t1b['TRUEZ']))
+        self.assertTrue(np.any(t1a['TRUEZ'] != t2['TRUEZ']))
+
+    #- Test that brighter moon makes noisier spectra
+    @unittest.skipIf('TRAVIS_JOB_ID' in os.environ, 'Skipping memory hungry quickbrick/specsim test on Travis')
+    def test_quickbrick_moon(self):
+        nspec = 2
+        cmd = "quickbrick --seed 1 -b brightmoon --objtype BRIGHT_MIX -n {} --outdir {} --moon-phase 0.1".format(nspec, self.testdir)
```
Comment (dkirkby) [inline py/desisim/test/test_quickbrick.py]: I would explicitly put the moon above the horizon here, instead of relying on a default.

File: py/desisim/test/test_quickbrick.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import numpy as np
+from astropy.io import fits
+from desisim.scripts import quickbrick
+import desispec.io
+
+class TestQuickBrick(unittest.TestCase):
+    
+    #- Pick random test directory
+    def setUp(self):
+        self.testdir = 'test-{uuid}'.format(uuid=uuid1())
```
Comment (dkirkby) [inline py/desisim/test/test_quickbrick.py]: It might be cleaner and more portable to use the [tempfile module](https://docs.python.org/2/library/tempfile.html) here.

File: py/desisim/test/test_quickbrick.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        nspec = 3
+        cmd = "quickbrick -b {} --objtype BGS -n {} --outdir {}".format(brickname, nspec, self.testdir)
+        cmd = cmd + " --airmass 1.5 --verbose --zrange-bgs 0.1 0.2"
+        cmd = cmd + " --moon-phase 0.1 --moon-angle 30 --moon-zenith 20"
+        args = quickbrick.parse(cmd.split()[1:])
+        results = quickbrick.main(args)
+        brickfile = '{}/brick-b-{}.fits'.format(self.testdir, brickname)
+        truth = fits.getdata(brickfile, '_TRUTH')        
+        z = truth['TRUEZ']
+        self.assertTrue(np.all((0.1 <= z) & (z <= 0.2)))
```
Comment (dkirkby) [inline py/desisim/test/test_quickbrick.py]: Could also check that the median ivar is within some tolerance of the expected value, to catch any changes that significantly change the results.

File: py/desisim/scripts/quickbrick.py [hunk 4]
```diff
@@ -228,8 +233,9 @@ def main(args):
     meta.add_column(Column(true_objtype, dtype=(str, 10), name='TRUE_OBJTYPE'))
     meta.add_column(Column(targetids, name='TARGETID'))
 
+    # @sbailey asks: What was this supposed to do for OBJTYPE vs. TRUE_OBJTYPE?
     # Now add fixed-up OBJTYPE to the template meta table
```
Comment (dkirkby) [inline py/desisim/scripts/quickbrick.py]: Hopefully @moustakas can answer this.

Comment (moustakas) [inline py/desisim/scripts/quickbrick.py]: Previously ```OBJTYPE``` was not included in the metadata table returned by ```desisim.templates```, but it was added as part of #132, which broke this part of the code.  We need this renaming of ```OBJTYPE--->TRUE_OBJTYPE``` for contaminants, however, for example, to identify objects with ```OBJTYPE='QSO_BAD'``` which are actually ```TRUE_OBJTYPE='STAR'```.  (Note that I didn't write this particular part of ```quickbrick``` but I believe that it's still needed downstream in the redshift fitting and classification portion of the pipeline.)

Long story short I think the line that's commented out can be removed safely.

**Comment by moustakas** [comment]
Modulo @dkirkby's comments, this looks fine to merge.

**Comment by sbailey** [comment]
Thanks for comments.
* I added options to ensure that moon is above horizon when testing `-b brightmoon` instead of relying upon defaults
* I added sanity check on median(error) vs. median(flux) for simulated spectra.  This test is intended to catch if something significantly changes; it isn't a detailed validation of the functional form of that.
* tempfile vs. uuid: we use this uuid method throughout multiple packages and tests.  I'm leaving as-is unless we find a specific problem with it that `tempfile` would fix.

I believe this addresses all comments and I will merge after tests pass on Travis.
